// nouislider-react/dist/nouislider-react.cjs.js
'use strict';

function _interopDefault (ex) { return (ex && (typeof ex === 'object') && 'default' in ex) ? ex['default'] : ex; }

var React = require('react');
var React__default = _interopDefault(React);
var nouislider = _interopDefault(require('nouislider'));

function _typeof(obj) {
  "@babel/helpers - typeof";

  if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") {
    _typeof = function (obj) {
      return typeof obj;
    };
  } else {
    _typeof = function (obj) {
      return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj;
    };
  }

  return _typeof(obj);
}

function _defineProperty(obj, key, value) {
  if (key in obj) {
    Object.defineProperty(obj, key, {
      value: value,
      enumerable: true,
      configurable: true,
      writable: true
    });
  } else {
    obj[key] = value;
  }

  return obj;
}

function _extends() {
  _extends = Object.assign || function (target) {
    for (var i = 1; i < arguments.length; i++) {
      var source = arguments[i];

      for (var key in source) {
        if (Object.prototype.hasOwnProperty.call(source, key)) {
          target[key] = source[key];
        }
      }
    }

    return target;
  };

  return _extends.apply(this, arguments);
}

function ownKeys(object, enumerableOnly) {
  var keys = Object.keys(object);

  if (Object.getOwnPropertySymbols) {
    var symbols = Object.getOwnPropertySymbols(object);
    if (enumerableOnly) symbols = symbols.filter(function (sym) {
      return Object.getOwnPropertyDescriptor(object, sym).enumerable;
    });
    keys.push.apply(keys, symbols);
  }

  return keys;
}

function _objectSpread2(target) {
  for (var i = 1; i < arguments.length; i++) {
    var source = arguments[i] != null ? arguments[i] : {};

    if (i % 2) {
      ownKeys(Object(source), true).forEach(function (key) {
        _defineProperty(target, key, source[key]);
      });
    } else if (Object.getOwnPropertyDescriptors) {
      Object.defineProperties(target, Object.getOwnPropertyDescriptors(source));
    } else {
      ownKeys(Object(source)).forEach(function (key) {
        Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key));
      });
    }
  }

  return target;
}

function _slicedToArray(arr, i) {
  return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _nonIterableRest();
}

function _arrayWithHoles(arr) {
  if (Array.isArray(arr)) return arr;
}

function _iterableToArrayLimit(arr, i) {
  if (!(Symbol.iterator in Object(arr) || Object.prototype.toString.call(arr) === "[object Arguments]")) {
    return;
  }

  var _arr = [];
  var _n = true;
  var _d = false;
  var _e = undefined;

  try {
    for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) {
      _arr.push(_s.value);

      if (i && _arr.length === i) break;
    }
  } catch (err) {
    _d = true;
    _e = err;
  } finally {
    try {
      if (!_n && _i["return"] != null) _i["return"]();
    } finally {
      if (_d) throw _e;
    }
  }

  return _arr;
}

function _nonIterableRest() {
  throw new TypeError("Invalid attempt to destructure non-iterable instance");
}

var sortObjectKeys = function sortObjectKeys(obj) {
  return Object.entries(obj).sort().reduce(function (o, _ref) {
    var _ref2 = _slicedToArray(_ref, 2),
        k = _ref2[0],
        v = _ref2[1];

    return o[k] = v, o;
  }, {});
};

var isEqual = function isEqual(val1, val2) {
  if (typeof val1 === "number" && typeof val2 === "number") return val1 === val2;
  if (typeof val1 === "string" && typeof val2 === "string") return val1 === val2;

  if (Array.isArray(val1) && Array.isArray(val2)) {
    return JSON.stringify(val1) === JSON.stringify(val2);
  }

  if (_typeof(val1) === "object" && _typeof(val2) === "object") {
    return JSON.stringify(sortObjectKeys(val1)) === JSON.stringify(sortObjectKeys(val2));
  }

  return false;
};

var areEqual = function areEqual(prevProps, nextProps) {
  var start = prevProps.start,
      step = prevProps.step,
      disabled = prevProps.disabled,
      range = prevProps.range;
  return nextProps.step === step && isEqual(nextProps.start, start) && nextProps.disabled === disabled && isEqual(nextProps.range, range);
};

var Nouislider = function Nouislider(props) {
  var _useState = React.useState(null),
      _useState2 = _slicedToArray(_useState, 2),
      slider = _useState2[0],
      setSlider = _useState2[1];

  var sliderContainer = React__default.createRef();
  React.useEffect(function () {
    var instanceRef = props.instanceRef;
    var isCreatedRef = instanceRef && Object.prototype.hasOwnProperty.call(instanceRef, "current");

    if (instanceRef && instanceRef instanceof Function) {
      instanceRef(sliderContainer.current);
    }

    if (isCreatedRef) {
      // eslint-disable-next-line no-param-reassign
      instanceRef.current = sliderContainer.current;
    }

    return function () {
      if (isCreatedRef) {
        // eslint-disable-next-line no-param-reassign
        instanceRef.current = null;
      }
    };
  }, [sliderContainer]);

  var clickOnPip = function clickOnPip(pip) {
    var value = Number(pip.target.getAttribute("data-value"));

    if (slider) {
      slider.set(value);
    }
  };

  var toggleDisable = function toggleDisable(disabled) {
    var sliderHTML = sliderContainer.current;

    if (sliderHTML) {
      if (!disabled) {
        sliderHTML.removeAttribute("disabled");
      } else {
        sliderHTML.setAttribute("disabled", true);
      }
    }
  };

  var updateOptions = function updateOptions(options) {
    var sliderHTML = sliderContainer.current;
    sliderHTML.noUiSlider.updateOptions(options);
  };

  var setClickableListeners = function setClickableListeners() {
    if (props.clickablePips) {
      var sliderHTML = sliderContainer.current;
      sliderHTML.querySelectorAll(".noUi-value").forEach(function (pip) {
        pip.style.cursor = "pointer";
        pip.addEventListener("click", clickOnPip);
      });
    }
  };

  var createSlider = function createSlider() {
    var onUpdate = props.onUpdate,
        onChange = props.onChange,
        onSlide = props.onSlide,
        onStart = props.onStart,
        onEnd = props.onEnd,
        onSet = props.onSet;
    var sliderComponent = nouislider.create(sliderContainer.current, _objectSpread2({}, props));

    if (onStart) {
      sliderComponent.on("start", onStart);
    }

    if (onSlide) {
      sliderComponent.on("slide", onSlide);
    }

    if (onUpdate) {
      sliderComponent.on("update", onUpdate);
    }

    if (onChange) {
      sliderComponent.on("change", onChange);
    }

    if (onSet) {
      sliderComponent.on("set", onSet);
    }

    if (onEnd) {
      sliderComponent.on("end", onEnd);
    }

    setSlider(sliderComponent);
  };

  React.useEffect(function () {
    var disabled = props.disabled;
    var sliderHTML = sliderContainer.current;

    if (sliderHTML) {
      toggleDisable(disabled);
      createSlider();
    }

    return function () {
      if (slider) slider.destroy();

      if (sliderHTML) {
        sliderHTML.querySelectorAll(".noUi-value").forEach(function (pip) {
          pip.removeEventListener("click", clickOnPip);
        });
      }
    };
  }, []);
  React.useEffect(function () {
    if (slider) {
      setClickableListeners();
    }
  }, [slider]);
  var start = props.start,
      disabled = props.disabled,
      range = props.range,
      step = props.step;
  React.useEffect(function () {
    if (slider) {
      updateOptions({
        range: range,
        step: step
      });
      slider.set(start);
      setClickableListeners();
    }

    toggleDisable(disabled);
  }, [start, disabled, range, step]);
  var id = props.id,
      className = props.className,
      style = props.style;
  var options = {};

  if (id) {
    options.id = id;
  }

  if (className) {
    options.className = className;
  }

  return React__default.createElement("div", _extends({}, options, {
    ref: sliderContainer,
    style: style
  }));
};

Nouislider.defaultProps = {
  animate: true,
  behaviour: "tap",
  className: null,
  clickablePips: false,
  connect: false,
  direction: "ltr",
  disabled: false,
  format: null,
  margin: null,
  limit: null,
  keyboardSupport: true,
  id: null,
  instanceRef: null,
  padding: 0,
  pips: null,
  snap: false,
  step: null,
  style: null,
  orientation: "horizontal",
  tooltips: false,
  onChange: function onChange() {},
  onEnd: function onEnd() {},
  onSet: function onSet() {},
  onSlide: function onSlide() {},
  onStart: function onStart() {},
  onUpdate: function onUpdate() {}
};
var index = React__default.memo(Nouislider, areEqual);

module.exports = index;
